components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1359,6 +1359,73 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
                           mojom::ProviderError::kSuccess, "");
 }
 
+void JsonRpcService::GetERC721TokenURI(const std::string& contract_address,
+                                       const std::string& token_id,
+                                       const std::string& chain_id,
+                                       GetERC721TokenURICallback callback) {
+  auto network_url = GetNetworkURL(prefs_, chain_id, mojom::CoinType::ETH);
+  if (!network_url.is_valid()) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  if (!EthAddress::IsValidAddress(contract_address)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  uint256_t token_id_uint = 0;
+  if (!HexValueToUint256(token_id, &token_id_uint)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  std::string data;
+  if (!erc721::TokenURI(token_id_uint, &data)) {
```

Comment:
AFAIK, metadata extension is optional for ERC721 smart contracts, what would happen if we try to call this on a ERC721 contract that does not support this method?

Reply:
I think we could just use https://github.com/brave/brave-core/blob/dd08110d888568002e0e03a31b267c6d83deee5f/components/brave_wallet/browser/json_rpc_service.cc#L1394 to check if this interface is supported.

Reply:
I was considering that, but wasn't sure it's strictly better.  This is my thinking:
* Calling supportsInterface adds at least one an additional RPC call for each NFT lookup.  It would be two calls if we check if the contract supports IERC165 first.
* I believe supportsInterface is optional and implemented by contract devs.  This has a few implications:
    * If the contract doesn't implement IERC165 / supportsInterface, then that eth_call would revert just as if it would revert if we called tokenURI on a contract that doesn't implement it
    * There could be contracts compliant with the ERC721 metadata extension, that do not implement IERC165 / supportsInterface. In that case our code wouldn't attempt to hit the `tokenURI` function even though it exists.
    * It's unlikely but possible for contracts to claim to implement the metadata extension when they don't.  In this case we'd encounter the revert anyway.

Reply:
To the second point, I think it's not optional cuz the ERC721 spec says "Every ERC-721 compliant contract must implement the ERC721 and ERC165 interfaces".

Reply:
Is there a document mentioning the fallback if defined will be non-payable?

Reply:
Seems like what fallback would do is not something well-defined and not we could control? If that's the case, supportsInterface seems better to me, wdyt?

Reply:
That's my understanding of fallback functions as well.

I guess my only question is about the potential for us to run into the same issue (hitting the fallback function and getting an error) when making the supportsInterface call.  If the contract does not implement IERC165, our call to the supportsInterface will hit the fallback if it exists and err / revert at that point. So I think we'd hit the fallback function through either approach for that case.

The way I'm thinking about it, the supportsInterface approach is superior when a contract implements IERC165, but does not implement the IERC721 Metadata extension.  In that case, we could call supportsInterface(IERC721MetaData) and get a boolean back instead of an error, and never call the fallback.  However, that's at the expensive of an additional eth_call.

Let me know if this changes your thinking at all.  I think either approach is workable, but the supportsInterface approach is slightly cleaner and feels more 'correct'.  I'd probably default to that if we are not concerned with the additional call.